File: src/ts/magma/app/compile/rule/NodeRule.ts
import { Node } from "../../../../magma/app/compile/node/Node";
import { Rule } from "../../../../magma/app/compile/rule/Rule";
import { Option } from "../../../../magma/api/option/Option";
import { MapNode } from "../../../../magma/app/compile/node/MapNode";
export class NodeRule implements Rule<Node> {
	key: string;
	rule: Rule<Node>;
	constructor (key: string, rule: Rule<Node>) {
		this.key = key;
		this.rule = rule;
	}
	lex(input: string): Option<Node> {
		return this.rule.lex(input).map(inner -  > new MapNode().withNode(this.key, inner))/*unknown*/;
	}
}

File: src/java/magma/app/compile/rule/NodeRule.java
package magma.app.compile.rule;

import magma.app.compile.node.MapNodeWithEverything;
import magma.app.compile.node.NodeWithEverything;

import java.util.Optional;

public record NodeRule(String key, Rule rule) implements Rule {
    @Override
    public Optional<String> generate(NodeWithEverything node) {
        return node.findNode(this.key)
                .flatMap(this.rule::generate);
    }

    @Override
    public Optional<NodeWithEverything> lex(String input) {
        return this.rule.lex(input)
                .map(node -> new MapNodeWithEverything().withNode(this.key, node));
    }
}
